Calculation and graph overhaul + Moon support

avataar opened the pull request:
Summary of changes:

- Completely revamped calculations and Sun graph, now way more realistic
- Moon support: Moon on the graph, moonrise, moonset, Moon phase, Moon elevation and Moon azimuth
- Obeys Home Assistant's settings for time and number formatting
- Cleaned up config options
- Font size adjustments, including the smaller AM/PM text from the original card
- Lots of tests
- All lint warnings gone

Resolves the following issues:
- #11
- #30
- #40
- #54
- #58

Thanks to @MiisaTrAnCe, @lcnittl, @edwardtfn, @4l4R1, @HydrelioxGitHub, @v1k70rk4, @lcnittl, @JasringStw, @tjorim, @HepoH3 and @v1k70rk4 for helping with translations for the new moonrise/moonset labels.

Note that I've also translated all existing languages where a contribution for the new labels was missing and I used Google Translate, plain old Google, Wikipedia, common sense and my own nerdiness for languages but some things might still be slightly off. I also corrected the sunrise/sunset labels for Malay as they were the labels for dusk/dawn instead (and I coincidentally learned that "matahari" means Sun in Malay and that's where Mata Hari got her stage name from!).

avataar commented:
Here is a built version for those who wish to test this but can't build it themselves (needs to be installed manually in HASS):
[lovelace-horizon-card.zip](https://github.com/rejuvenate/lovelace-horizon-card/files/11962341/lovelace-horizon-card.zip).

avataar commented:
Looking for brave souls who'd review/test this: @edwardtfn, @ThomDietrich, anyone!

HepoH3 commented:
> Looking for brave souls who'd review/test this: @edwardtfn, @ThomDietrich, anyone!

Thank you for you work!
Here is how it looks like to me (I played a bit with my PC time to get different results):
![изображение](https://github.com/rejuvenate/lovelace-horizon-card/assets/17159587/274d7e81-cb76-4f13-999c-cc381357a0cc)
![изображение](https://github.com/rejuvenate/lovelace-horizon-card/assets/17159587/16408c66-310f-40bf-8c5c-f4cba11ab511)
![изображение](https://github.com/rejuvenate/lovelace-horizon-card/assets/17159587/f719fac3-b3fb-4d9f-831e-a605f0379f03)

avataar commented:
@HepoH3, thanks, looks good! Check out the `now` option too, it will allow you to set a static time just for the card so you can get different looks without messing with the clock :) And of course the various fields options that show new Moon data.

HepoH3 commented:
> @HepoH3, thanks, looks good! Check out the `now` option too, it will allow you to set a static time just for the card so you can get different looks without messing with the clock :) And of course the various fields options that show new Moon data.

Moon fields looking good too (but I needed about five minutes to figure out that I'm missing `fields` option).
![изображение](https://github.com/rejuvenate/lovelace-horizon-card/assets/17159587/e53c10b6-a340-42f3-be5b-95ba7e6093bb)

The `now` option didn't work for me (I think I'm doing something wrong). I created a helper input_datetime object, and put it into `now field` which brokes everything.

```yaml
  - type: custom:horizon-card
    moon: true
    fields:
      moonrise: true
      moonset: true
      moon_phase: true
    # now: input_datetime.test_date
```


I think there should be an example of full card config.

JasringStw commented:
took me a moment to get the config right in HA but have it now, looks good. one comment is that maybe the title for the moon_phase should also be above the picture
![CleanShot 2023-07-06 at 11 00 49@2x](https://github.com/rejuvenate/lovelace-horizon-card/assets/44233200/7baf35ee-13e2-4b4f-a012-a289ea3db035)

ThomDietrich commented:
@JasringStw the position makes sense to me, as "afnemende maan" is the actual moon phase, therefore the state rather than the title!? However, because of that the text color would need to be black rather than gray @avataar ?

wwwescape commented:
Works great but I have one request which was from the older repo as well. Could we provide an option to have the field titles in title case?

![image](https://github.com/rejuvenate/lovelace-horizon-card/assets/651507/b37a43ec-54b4-4dac-bc02-d314c14eb8bd)

Perhaps something like: `useTitleCase`.

I could write up a helper function and create a PR once this is merged or maybe this PR could be updated with it. :grin:

avataar commented:
Thank you all for testing! I just pushed some fixes based on your comments:

- @HepoH3, the README now has an example config section showing the full config. The `now` should be something of the sort `2023-07-06T00:30:05+0300`
- @JasringStw and @ThomDietrich, I kept the Moon phase where it was but changed the colour to match the rest of the values
- @v1k70rk4, that was a tricky one but I figured it out. The Moon phase names come from Home Assistant's Moon integration and I'm pretty sure you don't have that enabled. I've changed the implementation to show the non-translated constant like `waning_gibbuous (!)` (and if you hover it with the mouse, a tooltip that explains the reason). I also added a Caveats section in the README that explains the peculiarities of Moon phase, including that it will be shown in HA's language and not the card language (if different). @HepoH3, your screenshots indicate the same problem.

I also removed the `dark_mode` option from the README as it's buggy - too many styles get inherited directly from Home Assistant. So you just get the same theme as your Home Assistant and I believe that's good enough.

New zip with the changes: [lovelace-horizon-card.zip](https://github.com/rejuvenate/lovelace-horizon-card/files/11971782/lovelace-horizon-card.zip)


@wwwescape, let's keep the title case thing for the near future after this is merged.

v1k70rk4 commented:
> * @v1k70rk4, that was a tricky one but I figured it out. The Moon phase names come from Home Assistant's Moon integration and I'm pretty sure you don't have that enabled. I've changed the implementation to show the non-translated constant like (and if you hover it with the mouse, a tooltip that explains the reason). I also added a Caveats section in the README that explains the peculiarities of Moon phase, including that it will be shown in HA's language and not the card language (if different). @HepoH3, your screenshots indicate the same problem.waning_gibbuous (!)

Thank you very much for your work. Indeed, the moon integration was not enabled. Now the colors and texts are perfect!

![image](https://github.com/rejuvenate/lovelace-horizon-card/assets/120401244/d292d2c4-5b03-45af-ad6b-570fcf93ee44)

wwwescape commented:
> Thank you all for testing! I just pushed some fixes based on your comments:
> 
> ...
>
> @wwwescape, let's keep the title case thing for the near future after this is merged.

@avataar No problem. I can wait. Thanks for all the work!

wwwescape commented:
> > Thank you all for testing! I just pushed some fixes based on your comments:
> > ...
> > @wwwescape, let's keep the title case thing for the near future after this is merged.
> 
> @avataar No problem. I can wait. Thanks for all the work!

We may not need the title case option after all! I managed to get the values title cased by style them using card-mod:

```
card_mod:
  style: |
    .horizon-card-field-name, .horizon-card-field-value {
      text-transform: capitalize;
    }
```

v1k70rk4 commented:
What needs to happen now for us to proceed? Or is there perhaps an issue? (I'm just asking out of curiosity :) )

avataar commented:
> What needs to happen now for us to proceed? Or is there perhaps an issue? (I'm just asking out of curiosity :) )

I've addressed all issues that were reported here + a fix for the recently opened issue #58 related to a new TZ setting in HA 2023.7.

Other than someone brave looking at the code there is nothing that prevents us from merging this :)

ThomDietrich commented on suncalc3/README-HORIZON-CARD.md:
```diff
@@ -0,0 +1,25 @@
+# Local copy of suncalc3
+
+This is a copy of module suncalc3 v2.0.5 with some minor modifications that bypass its wrong assumptions about time zones.
```
Seems like an extreme step. Did you raise those wrong assumptions, do you see a path to remove this local copy at a later state?

edwardtfn replied:
"its incorrect assumptions", or "imprecise", sounds better than "wrong". 😉
"some minor modifications to improve its time zone handling."?

avataar replied:
@ThomDietrich, extreme or not it was very necessary. When I find the time I might submit a PR to the original project but I really needed it to work in an "improved" way :)

@edwardtfn, wrong is wrong whether you call it that or "incorrect". But I get your point so I changed it to "improve its time zone handling".

ThomDietrich reviewed (approved):
Amazing work!

ThomDietrich commented:
@edwardtfn would you like to leave a review?

This is an amazing addition to the project and I am more than happy to support its merge. @avataar please proceed with the merge, tagging, versioning and release notes as you see fit :)

edwardtfn commented on README.md:
```diff
@@ -140,6 +180,43 @@ Supported options for the `language` setting:
-
-- Home Assistant reports the time of the next occurring Sun event. For example, if you look at the card during the day, the time for sunrise will reflect tomorrow's sunrise and not the one that occurred on the same day.
+### Caveats
+
```
"... followed by `(!)`", right?

avataar replied:
Good spot! Fixed.
